builtin/receive-pack.c

```diff
@@ -997,16 +998,26 @@ static const char *push_to_checkout(unsigned char *hash,
 		return NULL;
 }
 
-static const char *update_worktree(unsigned char *sha1)
+static const char *update_worktree(unsigned char *sha1, const struct worktree *worktree)
 {
 	const char *retval;
-	const char *work_tree = git_work_tree_cfg ? git_work_tree_cfg : "..";
+	const char *work_tree;
 	struct argv_array env = ARGV_ARRAY_INIT;
 
+	if (worktree->path)
+		work_tree = worktree->path;
+	else if (git_work_tree_cfg)
+		work_tree = git_work_tree_cfg;
+	else
+		work_tree = "..";
+
+	//const char *work_tree = git_work_tree_cfg ? git_work_tree_cfg : "..";
+
 	if (is_bare_repository())
 		return "denyCurrentBranch = updateInstead needs a worktree";
 
-	argv_array_pushf(&env, "GIT_DIR=%s", absolute_path(get_git_dir()));
+	argv_array_pushf(&env, "GIT_DIR=%s", absolute_path(get_worktree_git_dir(worktree)?get_worktree_git_dir(worktree):get_git_dir()));
```

Comment:
It would be better to initialize a variable instead:

```c
const char *git_dir = NULL;
[...]
if (worktree)
    git_dir = get_worktree_git_dir(worktree);
if (!git_dir)
    git_dir = get_git_dir();
argv_array_pushf(&env, "GIT_DIR=%s", absolute_path(git_dir));
```